share/source.F90

```diff
@@ -72,24 +81,36 @@ subroutine source
    rands(4) = rang() ! sample y
    rands(5) = rang() ! sample z
  
-   cell_num = -1
-   call particle_birth(rands, xxx, yyy, zzz, erg, wgt, cell_num)
-   icl_tmp = find_cell()
-   if (idum(1) > 2) then
-       if (cell_num .ne. ncl(icl_tmp) .and. tries < idum(2)) then
-           tries = tries + 1
-           goto 200
-       endif
+   call particle_birth(rands, xxx, yyy, zzz, erg, wgt, cell_list)
+   ! In tet mesh, cell_list_size = 0, loop over entire cells to find icl_tmp
+   if (cell_list_size == 0) then
+       icl_tmp = find_cell(ncl, mxa)
+   else
+   ! In Cartisian mesh (voxel/sub-voxel), loop over cell_list to find icl_tmp
+       icl_tmp = find_cell(cell_list, cell_list_size)
    endif
-   if (mat(icl_tmp).eq.0 .and. tries < idum(2)) then
-       tries = tries + 1
-       goto 200
-   end if
 
-   if(tries.eq.idum(2)) then
+   ! check whether this is a valid cell
+   if (icl_tmp .le. 0) then
```

Comment:
When we use `cell_list` to find out the cell, there is a very small possibility to find that the `xxx, yyy, zzz` located in a place not listed in the `cell_list` (because ray tracing may not find that small cell in the voxel). In that case, the `icl_tmp` is set to `-1`. And then sample the particle again.

Reply:
Sounds good. Thank you for clarifying!